airbyte-integrations/bases/base-java-s3/src/test/java/io/airbyte/integrations/destination/s3/S3BaseChecksTest.java

```diff
@@ -0,0 +1,49 @@
+package io.airbyte.integrations.destination.s3;
+
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.ArgumentMatchers.matches;
+import static org.mockito.ArgumentMatchers.startsWith;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.model.ListObjectsRequest;
+import io.airbyte.integrations.destination.s3.util.S3NameTransformer;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.ArgumentMatchers;
+
+public class S3BaseChecksTest {
+
+  private AmazonS3 s3Client;
+
+  @BeforeEach
+  public void setup() {
+    s3Client = mock(AmazonS3.class);
+  }
+
+  @Test
+  public void attemptWriteAndDeleteS3Object_should_createSpecificFiles() {
+    S3DestinationConfig config = new S3DestinationConfig(
+        null,
+        "test_bucket",
+        "test/bucket/path",
+        null,
+        null,
+        null,
+        null,
+        s3Client
+    );
+    S3StorageOperations operations = new S3StorageOperations(new S3NameTransformer(), s3Client, config);
+    when(s3Client.doesObjectExist("test_bucket", "test/bucket/path/")).thenReturn(false);
+
+    S3BaseChecks.attemptS3WriteAndDelete(operations, config, "test/bucket/path");
+
+    verify(s3Client).putObject("test_bucket", "test/bucket/path/", "");
+    verify(s3Client).putObject(eq("test_bucket"), startsWith("test/bucket/path/_airbyte_connection_test_"), anyString());
+    verify(s3Client).listObjects(ArgumentMatchers.<ListObjectsRequest>argThat(request -> "test_bucket".equals(request.getBucketName())));
+    verify(s3Client).deleteObject(eq("test_bucket"), startsWith("test/bucket/path/_airbyte_connection_test_"));
```

Comment:
@N-o-Z I went ahead and put together a quick unit test so we don't have another regression here - does this look right to you?

Reply:
Not sure, as far as I understand attemptS3WriteAndDelete performs more or less the same think you are testing afterwards (create, list, delete on the path). If I understand correctly, this test runs the same scenario twice.

Also, please note that this does not prevent recurrence of the original problem. To do that you need to have a test where write permissions are only allowed on the bucket path itself and not on the bucket root and ensure we can create a new S3 destination connector successfully using this configuration.

Reply:
interesting. There's probably some refactoring to do in this part of the codebase later - certainly `attemptS3WriteAndDelete` is getting called without the subsequent put/list/delete in other connectors, e.g.  [destination-redshift](https://github.com/airbytehq/airbyte/blob/cc41385f507d8711445bbb713e98872741caf19d/airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftStagingS3Destination.java#L70)

I think the new unit test (somewhat indirectly) verifies that original problem? Or at least it enforces that we're not doing `putObject("test_bucket", "test_bucket")`... regardless, filed https://github.com/airbytehq/airbyte/issues/18438 because I agree we should verify the end to end behavior

Reply:
Sounds good. The issue you've opened + the new unit test will certainly cover all case. Thanks a lot